# **HR Attrition - Introduction and Methodology**
Group 10:
- Filipe Brandão Carmo
- João Silva
- Rita Marques
- Sara Henriques

The objective of this capstone project is to synthesize and apply the diverse skills and knowledge you have acquired throughout your coursework. You will
develop an end-to-end data science solution focused on predicting employee attrition. The project is designed to challenge you to integrate your
understanding of data exploration, feature engineering, machine learning, and model deployment. Additionally, you will refine your ability to communicate
complex results through a polished, professional presentation.

GitHub Repository: Maintain a well-organized GitHub repository containing all project files, including code, data, documentation, and the slide
deck. Ensure the repository is structured and commented to be understandable by others.

Deliverables:
1. Jupyter Notebook/Python Scripts: Detailed documentation of your entire
workflow, from EDA to modeling and deployment.
2. Final Presentation: A polished, professional presentation with
accompanying slides, ready for a 10-minute delivery.
3. Backup Slides: Additional slides to address potential questions or provide
further details.
4. GitHub Repository: A complete and organized repository with all relevant
files and documentation.

# **Part I. Understanding the Business Problem**

High turnover rates are costly and disruptive, making it essential for HR to anticipate which employees are likely to leave. We are tasked with assisting a multinational consultancy firm in predicting employee attrition.

The primary goal is to predict whether an employee will leave the company, based on the data provided. Secondary goals include identifying key factors influencing attrition and recommending strategies to retain valuable employees.

The HR department and executive management will be the primary consumers of your insights, and they expect actionable recommendations based on your analysis.

# **Part II - Data Integration**

The dataset provided (HR_DS.csv) contains various attributes related to employee demographics, job satisfaction, work experience, and compensation.

Explore the data types, distribution, and completeness.
Identify any missing values or inconsistencies that need to be addressed.´

Preprocessing:
- Handle missing data through appropriate imputation techniques.
- Encode categorical variables using methods such as one-hot encoding or label encoding, considering the impact on the model.
- Standardize or normalize numerical features if necessary, based on the algorithms you plan to use.

Technologies to be used and Python packages to import:
- Programming & Data Analysis: Python (Pandas, NumPy, Scikit-learn, XGBoost, TensorFlow/PyTorch, etc.)
- Development Environment: VS Code
- Version Control: Git/GitHub
- Model Deployment: Hugging Face Spaces (Optional)
- Data Visualization: Power BI, Plotly, Matplotlib, Seaborn
- Model Interpretation: SHAP, LIME

### **1. Import the needed libraries and dataset**

In [23]:
# imports
import numpy as np
import pandas as pd
import seaborn as sns

In [24]:
# import data here and perform basic .shape() .info()
HR = pd.read_csv('HR_Attrition_Dataset.csv')
print(f"The table has the following format: ", HR.shape)
print(HR.info(max_cols=35))

The table has the following format:  (1470, 35)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470

### **2. Set index**

A quick exploration in the dataset allowed to undestand that EmployeeNumber had the potential to be used as index and since those values are unique.

In [25]:
pd.set_option('display.max_columns', None)
HR.head(5)

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   EnvironmentSatisfaction  Gender  HourlyRate  JobInvolvement  JobLevel  \
0                        2  Female          94               3         2   
1                        3    Male          61               2         2   
2                        4    Male          92               2         1   
3                        4  Female          56               3         1   
4                        1    Male          40               3         1   

                 JobRole  JobSatisfaction MaritalStatus  MonthlyIncome  \
0        Sales Executive                4        Single           5993   
1     Research Scientist                2       Married           5130   
2  Laboratory Technician                3        Single           2090   
3     Research Scientist                3       Married           2909   
4  Laboratory Technician                2       Married           3468   

   MonthlyRate  NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
0        19479                   8      Y      Yes                 11   
1        24907                   1      Y       No                 23   
2         2396                   6      Y      Yes                 15   
3        23159                   1      Y      Yes                 11   
4        16632                   9      Y       No                 12   

   PerformanceRating  RelationshipSatisfaction  StandardHours  \
0                  3                         1             80   
1                  4                         4             80   
2                  3                         2             80   
3                  3                         3             80   
4                  3                         4             80   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                 0                  8                      0   
1                 1                 10                      3   
2                 0                  7                      3   
3                 0                  8                      3   
4                 1                  6                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                1               6                   4   
1                3              10                   7   
2                3               0                   0   
3                3               8                   7   
4                3               2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2

In [26]:
unique_values = HR['EmployeeNumber'].is_unique
print (unique_values)

True


In [27]:
HR.set_index('EmployeeNumber', inplace = True)
HR.head(5)

Age Attrition     BusinessTravel  DailyRate  \
EmployeeNumber                                                
1                41       Yes      Travel_Rarely       1102   
2                49        No  Travel_Frequently        279   
4                37       Yes      Travel_Rarely       1373   
5                33        No  Travel_Frequently       1392   
7                27        No      Travel_Rarely        591   

                            Department  DistanceFromHome  Education  \
EmployeeNumber                                                        
1                                Sales                 1          2   
2               Research & Development                 8          1   
4               Research & Development                 2          2   
5               Research & Development                 3          4   
7               Research & Development                 2          1   

               EducationField  EmployeeCount  EnvironmentSatisfaction  Gender  \
EmployeeNumber                                                                  
1               Life Sciences              1                        2  Female   
2               Life Sciences              1                        3    Male   
4                       Other              1                        4    Male   
5               Life Sciences              1                        4  Female   
7                     Medical              1                        1    Male   

                HourlyRate  JobInvolvement  JobLevel                JobRole  \
EmployeeNumber                                                                
1                       94               3         2        Sales Executive   
2                       61               2         2     Research Scientist   
4                       92               2         1  Laboratory Technician   
5                       56               3         1     Research Scientist   
7                       40               3         1  Laboratory Technician   

                JobSatisfaction MaritalStatus  MonthlyIncome  MonthlyRate  \
EmployeeNumber                                                              
1                             4        Single           5993        19479   
2                             2       Married           5130        24907   
4                             3        Single           2090         2396   
5                             3       Married           2909        23159   
7                             2       Married           3468        16632   

                NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
EmployeeNumber                                                          
1                                8      Y      Yes                 11   
2                                1      Y       No                 23   
4                                6      Y      Yes                 15   
5                                1      Y      Yes                 11   
7                                9      Y       No                 12   

                PerformanceRating  RelationshipSatisfaction  StandardHours  \
EmployeeNumber                                                               
1                               3                         1             80   
2                               4                         4             80   
4                               3                         2             80   
5                               3                         3             80   
7                               3                         4             80   

                StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
EmployeeNumber                                                               
1                              0                  8                      0   
2                              1                 10                      3   
4                              0           

### **3. Check duplicates**

The command below was used to find duplicated lines - any register was found on those conditions.

In [28]:
HR[HR.duplicated()]

Empty DataFrame
Columns: [Age, Attrition, BusinessTravel, DailyRate, Department, DistanceFromHome, Education, EducationField, EmployeeCount, EnvironmentSatisfaction, Gender, HourlyRate, JobInvolvement, JobLevel, JobRole, JobSatisfaction, MaritalStatus, MonthlyIncome, MonthlyRate, NumCompaniesWorked, Over18, OverTime, PercentSalaryHike, PerformanceRating, RelationshipSatisfaction, StandardHours, StockOptionLevel, TotalWorkingYears, TrainingTimesLastYear, WorkLifeBalance, YearsAtCompany, YearsInCurrentRole, YearsSinceLastPromotion, YearsWithCurrManager]
Index: []

# **Part III. Exploratory Data Analysis**

- Univariate Analysis: Conduct a thorough analysis of individual variables to understand their distributions and detect any anomalies or patterns.
- Bivariate and Multivariate Analysis: Examine the relationships between key features, particularly how they correlate with the target variable (Attrition). Utilize heatmaps, pair plots, and correlation matrices to uncover potential multicollinearity issues.
- Visualization: Develop insightful visualizations to communicate findings effectively. Consider using tools like Seaborn or Plotly to create interactive or
complex visualizations.
- Feature Engineering: Assess feature importance using techniques like feature selection or model-based importance scoring (e.g., Random Forest feature importances).

## **1. Info & Shape**

The dataset has a total of 1470 lines with 34 attributes.

In [29]:
HR.shape

(1470, 34)

In [30]:
HR.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField', 'EmployeeCount',
       'EnvironmentSatisfaction', 'Gender', 'HourlyRate', 'JobInvolvement',
       'JobLevel', 'JobRole', 'JobSatisfaction', 'MaritalStatus',
       'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked', 'Over18',
       'OverTime', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StandardHours', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

In [31]:
pd.set_option('display.max_columns', None)
HR.head(5)

Age Attrition     BusinessTravel  DailyRate  \
EmployeeNumber                                                
1                41       Yes      Travel_Rarely       1102   
2                49        No  Travel_Frequently        279   
4                37       Yes      Travel_Rarely       1373   
5                33        No  Travel_Frequently       1392   
7                27        No      Travel_Rarely        591   

                            Department  DistanceFromHome  Education  \
EmployeeNumber                                                        
1                                Sales                 1          2   
2               Research & Development                 8          1   
4               Research & Development                 2          2   
5               Research & Development                 3          4   
7               Research & Development                 2          1   

               EducationField  EmployeeCount  EnvironmentSatisfaction  Gender  \
EmployeeNumber                                                                  
1               Life Sciences              1                        2  Female   
2               Life Sciences              1                        3    Male   
4                       Other              1                        4    Male   
5               Life Sciences              1                        4  Female   
7                     Medical              1                        1    Male   

                HourlyRate  JobInvolvement  JobLevel                JobRole  \
EmployeeNumber                                                                
1                       94               3         2        Sales Executive   
2                       61               2         2     Research Scientist   
4                       92               2         1  Laboratory Technician   
5                       56               3         1     Research Scientist   
7                       40               3         1  Laboratory Technician   

                JobSatisfaction MaritalStatus  MonthlyIncome  MonthlyRate  \
EmployeeNumber                                                              
1                             4        Single           5993        19479   
2                             2       Married           5130        24907   
4                             3        Single           2090         2396   
5                             3       Married           2909        23159   
7                             2       Married           3468        16632   

                NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
EmployeeNumber                                                          
1                                8      Y      Yes                 11   
2                                1      Y       No                 23   
4                                6      Y      Yes                 15   
5                                1      Y      Yes                 11   
7                                9      Y       No                 12   

                PerformanceRating  RelationshipSatisfaction  StandardHours  \
EmployeeNumber                                                               
1                               3                         1             80   
2                               4                         4             80   
4                               3                         2             80   
5                               3                         3             80   
7                               3                         4             80   

                StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
EmployeeNumber                                                               
1                              0                  8                      0   
2                              1                 10                      3   
4                              0           

In [ ]:
# perform .describe() and histograms/variable for univariate data analysis

In [ ]:
# Multivariate Analysis: correlation heatmap .corr()

# **Part IV. Predictive Modelling**

Model Selection:
- Start with basic models (e.g., Logistic Regression) to establish a performance baseline.
- Experiment with more complex models such as Decision Trees, Random Forests, Gradient Boosting Machines, and Neural Networks.

Model Tuning:
- Perform hyperparameter tuning using Grid Search or Random Search with cross-validation to optimize model performance.
- Address overfitting through techniques such as cross-validation, regularization, or pruning.

Model Evaluation:
- Evaluate models using a comprehensive set of metrics: accuracy, precision, recall, F1-score, ROC-AUC, and confusion matrix.
- Perform a detailed error analysis to understand the model’s strengths and weaknesses, particularly in predicting minority classes (e.g., those at high risk
of attrition).

Model Interpretation:
- Use techniques like SHAP (SHapley Additive exPlanations) or LIME (Local Interpretable Model-agnostic Explanations) to interpret complex models and explain predictions to non-technical stakeholders.

In [ ]:
# oof, até tenho pesadelos com isto

In [ ]:
# subset HR into train, test and validation (80, 10, 10?)

# 5. Prescriptive Analytics and Recommendations

- Insights Derivation: Based on the model’s predictions, derive actionable insights. For example, identify the top factors leading to attrition and propose
specific interventions.
- Scenario Analysis: Conduct what-if scenarios to simulate the impact of different interventions on attrition rates.
- Strategic Recommendations: Provide a set of clear, evidence-based recommendations for HR and management. For example, suggest targeted retention strategies such as personalized development plans for at-risk employees, or changes in compensation structures.

In [ ]:
# Maybe we don't need code here

# 6. Model Deployment (Optional)

Deploy the best-performing model using Hugging Face Spaces or a similar platform.

# 7. Final Presentation and Reporting

Presentation Structure:
- Introduction: Briefly introduce the business problem, objectives, and approach.
- EDA and Feature Engineering: Summarize key findings from your data exploration and the rationale behind your feature engineering choices.
- Modeling and Evaluation: Discuss the models you developed, their performance metrics, and why the final model was selected.
- Recommendations: Present your strategic recommendations based on the model’s insights.

Slide Deck:
- Create a professional slide deck with a logical flow, clear visuals, and concise explanations.
- Prepare backup slides with additional data and analysis for potential questions or deep